src/models/helpers.rs

```diff
@@ -0,0 +1,256 @@
+//! Supporting Functions and Types (VirtualTargetPath, safe_path)
+use serde::de::{Deserialize, Deserializer, Error as DeserializeError};
+use std::collections::{HashMap, HashSet};
+use std::fmt::{Debug};
+use std::str;
+
+use serde_derive::{ Serialize};
+
+use crate::crypto::{HashAlgorithm, HashValue};
+use crate::Result;
+
+use crate::error::Error;
+
+
+#[rustfmt::skip]
+static PATH_ILLEGAL_COMPONENTS: &'static [&str] = &[
+    ".", // current dir
+    "..", // parent dir
+         // TODO ? "0", // may translate to nul in windows
+];
+
+#[rustfmt::skip]
+static PATH_ILLEGAL_COMPONENTS_CASE_INSENSITIVE: &'static [&str] = &[
+    // DOS device files
+    "CON",
+    "PRN",
+    "AUX",
+    "NUL",
+    "COM1",
+    "COM2",
+    "COM3",
+    "COM4",
+    "COM5",
+    "COM6",
+    "COM7",
+    "COM8",
+    "COM9",
+    "LPT1",
+    "LPT2",
+    "LPT3",
+    "LPT4",
+    "LPT5",
+    "LPT6",
+    "LPT7",
+    "LPT8",
+    "LPT9",
+    "KEYBD$",
+    "CLOCK$",
+    "SCREEN$",
+    "$IDLE$",
+    "CONFIG$",
+];
+
+#[rustfmt::skip]
+static PATH_ILLEGAL_STRINGS: &'static [&str] = &[
+    ":", // for *nix compatibility
+    "\\", // for windows compatibility
+    "<",
+    ">",
+    "\"",
+    "|",
+    "?",
+    "*",
+    // control characters, all illegal in FAT
+    "\u{000}",
+    "\u{001}",
+    "\u{002}",
+    "\u{003}",
+    "\u{004}",
+    "\u{005}",
+    "\u{006}",
+    "\u{007}",
+    "\u{008}",
+    "\u{009}",
+    "\u{00a}",
+    "\u{00b}",
+    "\u{00c}",
+    "\u{00d}",
+    "\u{00e}",
+    "\u{00f}",
+    "\u{010}",
+    "\u{011}",
+    "\u{012}",
+    "\u{013}",
+    "\u{014}",
+    "\u{015}",
+    "\u{016}",
+    "\u{017}",
+    "\u{018}",
+    "\u{019}",
+    "\u{01a}",
+    "\u{01b}",
+    "\u{01c}",
+    "\u{01d}",
+    "\u{01e}",
+    "\u{01f}",
+    "\u{07f}",
+];
+
+pub fn safe_path(path: &str) -> Result<()> {
```

Comment:
e.g., I wonder if instead of dealing with string slices here we can manage things using std::path. This may be something we can address as a standalone issue.

Reply:
Gotcha, so leaving the path handling to std::path since it's more complete